### ISSUE TITLE
Potential locking problems due to the inconsistent checking

### ISSUE THREAD

**Issue by ycaibb** [openssl/openssl#17353]
Hi, I found a potential locking problem.

When `category < 0` at line 466, the lock `trace_lock`at Line 473 is not acquired.
https://github.com/openssl/openssl/blob/a935791d54078f43209ffbc1886ac5e68772ce34/crypto/trace.c#L459-L473

The range of `category` is not checked at line 498, lock `trace_lock` could be released without acquiring at line 516?
https://github.com/openssl/openssl/blob/a935791d54078f43209ffbc1886ac5e68772ce34/crypto/trace.c#L493-L516

**Comment by t-j-h** [comment]
The checks of category range are horribly inconsistent here ... and they should all be the same.

The consistent test for not-valid category should be:

`    if (category < 0 || category >= OSSL_TRACE_CATEGORY_NUM)
`

And for valid should be:

`    if (category >= 0 && category < OSSL_TRACE_CATEGORY_NUM)`

**Comment by paulidale** [comment]
Definitely a bug.

**Comment by mattcaswell** [comment]
Yes a bug that should be fixed, and the range checks should be consistent as per @t-j-h's comment - but the impact of this bug is minimal/negligible. If `category < 0` then `OSSL_trace_begin()` will return NULL. If the caller is using it in accordance with the documentation then `OSSL_trace_end()` should be passed the BIO returned from `OSSL_trace_begin()`. Since that is NULL, it will never hit the unlock code in `OSSL_trace_end()`.

If the caller is using the convenience macros `OSSL_TRACE_BEGIN` and `OSSL_TRACE_END` correctly (as is the case through the OpenSSL code internally) then this bug will never be hit.

**Comment by ycaibb** [comment]
Thank you very much for your checking.

**Comment by mattcaswell** [comment]
> The checks of category range are horribly inconsistent here ... and they should all be the same.

Actually I looked further at this. The checks are consistent already. Where we have code that looks like this:

```
    category = ossl_trace_get_category(category);
    if (category < 0)
        return NULL;
````
The full range check is included as part of `ossl_trace_get_category()`. If it is out of the range (`category < 0 || category >= OSSL_TRACE_CATEGORY_NUM`) then the return value from `ossl_trace_get_category()` is always -1, so a check of `category < 0` is sufficient, and anything more would be redundant.

The only thing missing here is a check for `category < 0` in `OSSL_trace_end()`

**Comment by mattcaswell** [comment]
Fix in #17371
